### DISCUSSION TITLE
ByteArrayType for Entity Framework RowVersion/[TimeStamp]

### DISCUSSION THREAD

**Discussion by OneLogicalMyth** [ChilliCream/hotchocolate#3132] [Q&A] [answered]
Hi,

Further question I can't workout but hopefully you can explain it for me. I use row version on my objects to track concurrency issues.

~~~c#
public class MyObject
{
    public int Id { get; set; }
    public string Name { get; set; }

    [TimeStamp]
    public byte[] RowVersion { get; set; }
}
~~~

I just worked out how to register/add the `ByteArrayType`, I have done so like this in startup.cs;

~~~c#
 services
    .AddGraphQLServer()
    // ... cut short for brevity
    .AddType(new UuidType('D'))
    .AddType<ByteArrayType>();
~~~

I can see looking at the source this should make a nice base64 string which is how it would work if I implemented a controller REST API equivalent so this is great.

However, I get an error which seems to suggest its only looking at the single byte (because it says `System.Byte` without [] maybe I am wrong):

~~~json
  "errors": [
    {
      "message": "ByteArray cannot serialize the given value.",
      "locations": [
        {
          "line": 29,
          "column": 5
        }
      ],
      "path": [
        "myObjects",
        19,
        "rowVersion",
        0
      ],
      "extensions": {
        "actualValue": "0",
        "actualType": "System.Byte",
        "code": "EXEC_INVALID_LEAF_VALUE"
      }
    },
~~~

I am pulling the data from my `DbSet<MyObject>` DB context, which is then returned as an IQueryable.

In my code-first type I am declaring the descriptor description and nothing else:

~~~c#
            descriptor
                .Field(v => v.RowVersion)
                .Description("Row version used for comparing on updates or deletions.");
~~~

For my GUID the format I wanted was 'D' as it helps our other applications work with the endpoint. I didn't have to add anything to the descriptor to make this work. But maybe I have to for a byte array to work?

Any ideas what I am missing? 

thanks

**Comment by michaelstaib** [comment]
Can you create a small repro?

**Reply by OneLogicalMyth** [reply]
Good idea, I will make one over the weekend.

**Reply by OneLogicalMyth** [reply]
Here you go, I made a new project and was able to reproduce the error. Pretty sure its myself and not a bug:

https://github.com/OneLogicalMyth/ByteArrayTest

**Reply by OneLogicalMyth** [reply]
![image](https://user-images.githubusercontent.com/12527543/109416232-d637fe00-79b4-11eb-81ad-d7cf9353d786.png)

**Reply by OneLogicalMyth** [reply]
For further reference I am using the following Docker container;

* mcr.microsoft.com/mssql/server:2017-latest

The table created has the following structure within the databse:

~~~sql
CREATE TABLE [dbo].[Objects](
	[Id] [uniqueidentifier] NOT NULL,
	[Name] [nvarchar](450) NULL,
	[RowVersion] [timestamp] NULL,
 CONSTRAINT [PK_Objects] PRIMARY KEY CLUSTERED 
(
	[Id] ASC
)WITH (PAD_INDEX = OFF, STATISTICS_NORECOMPUTE = OFF, IGNORE_DUP_KEY = OFF, ALLOW_ROW_LOCKS = ON, ALLOW_PAGE_LOCKS = ON) ON [PRIMARY]
) ON [PRIMARY]
GO
~~~

**Reply by michaelstaib** [reply]
Thanks for putting this together ... I will have a look tomorrow.

**Reply by eduardomhg** [reply]
Any news or work arounds on this? I See the same issue on version 12 RC when I add this to the config:

`.BindRuntimeType<byte[], ByteArrayType>();`

And I see it for any defined byte array, such as:

`public byte[] Test => new byte[]{ 1, 2, 3};`

**Comment by gojanpaolo** [comment, marked as answer]
Explicitly configuring the `RowVersion` field to be of `Type<ByteArrayType>` should resolve the error:

```c#
public class MyObjectObjectType : ObjectType<MyObject>
{
    protected override void Configure(IObjectTypeDescriptor<MyObject> descriptor)
    {
        descriptor
            .Field(_ => _.RowVersion)
            .Type<ByteArrayType>();
    }
}

// in Startup ConfigureServices
services
    .AddGraphQLServer()
    .AddType<MyObjectObjectType>()
    ...
```

It may be a bug on HC implicit configuration for `byte[]` when `ByteArrayType` is added.

**Reply by OneLogicalMyth** [reply]
This did the trick thank you!

Could be a bug then as when I used `UuidType` on the services this applied to all of them implicitly.

**Reply by michaelstaib** [reply]
This seems to be a bug... just looked at the code. Can you raise an issue.

Thanks for your help @gojanpaolo

**Reply by JoepLijnen** [reply]
I just ran into this very bug again. Was a bug registered for this? It also seems very, very similar to what is required to get NetTopologySuite types to work. This is leading to over 200+ lines of boiler plate in our application. (there it entails specifying sortinputtyhpes instead of here, where we will need objecttypes and registering them.).

**Comment by eduardogoncalves** [comment]
@gojanpaolo how to make it work implicity?
